### PR TITLE
Apply formatters and aggregators properly

### LINKED ISSUE #4444: -to-many relationships are not being aggregated
-to-many relationships are not being aggregated when a table format references them (for example, `determinations` or `preparations` do not display aggregated when added to a table format).
Recursive formatters are also not limited, but once aggregation can be displayed, this must be handled to prevent future errors.

### PR DESCRIPTION
**Pull request by realVinayak**
Fixes #4444 

Testing instructions:
Multiple cases that need to be tested, so I'll recommend doing them incrementally. I'll add the test cases for all of them soon, but here are they for the first case:

**To-many not being aggregated when referenced in a formatter**
1. Open record formatter visual editor for a table.
2. Add a field which is to-many (for example, Determinations for Collection Object). It would appear as (aggregated)
3. If using Determinations for Collection Object, make a new collection object with more than 1 Determination.
4. Query that collection object, using (formatted) as the field. You should see that both of the determinations have been included in a single line.
Example:
image
My collection object has two determinations, one with Taxon named "Abantennarius will be added" and other with name "Aborichthys aa" and they both appear (since formatter of my determination is taxon)
image

### CODE REVIEW

File: specifyweb/stored_queries/format.py
Comment (maxxxxxdlp) [inline specifyweb/stored_queries/format.py:182]: huh, didn't know python allowed that! I though it required return to be on a new line

File: specifyweb/stored_queries/format.py
Comment (maxxxxxdlp) [inline specifyweb/stored_queries/format.py:188]: weird formatting here. can you make each attribute start on a new line, just like prettier would do that in javascript?

**Review by maxxxxxdlp — APPROVED** [review]
Not too knowledgeable on python code. I will defer to Jason for a deeper review

File: specifyweb/stored_queries/format.py
Comment (realVinayak) [reply specifyweb/stored_queries/format.py:182]: Just some weird things Python allows you to shoot your foot with

File: specifyweb/stored_queries/format.py
Comment (melton-jason) [reply specifyweb/stored_queries/format.py:188]: I'm not entirely sure, but I believe @realVinayak formatted the document with PEP 8 (the Python standard for formatting and style). 
If that is the case, then this change was done 'automatically'. 

@realVinayak which formatter did you use to format the document with? 
Or was it all done by hand? :grin:

Comment (realVinayak) [reply specifyweb/stored_queries/format.py:188]: Lol it was done by PEP 8. I can revert the formatting if that's a problem.

**Comment by chanulee1** [comment]
How do I do UX testing for this?

File: specifyweb/stored_queries/format.py
Comment (maxxxxxdlp) [reply specifyweb/stored_queries/format.py:188]: Not a problem, just very strange - i.e, why does it have 2 arguments on first line and then 2 arguments on the next line rather than one argument per line (like in js) - that seems far more readable

Probably I'm just not used to the intricacies of the python world

File: specifyweb/stored_queries/format.py
Comment (maxxxxxdlp) [inline specifyweb/stored_queries/format.py:124]: this looks like the same algorithm front-end uses for cycle detection 👍:

https://github.com/specify/specify7/blob/a4a66e63a8ec27c82c60bca0c01bc0e1410b0725/specifyweb/frontend/js_src/lib/components/Formatters/formatters.ts#L147-L154C9

**Review by maxxxxxdlp — APPROVED** [review]

**Comment by pashiav** [comment]
Instructions for testers? @realVinayak

File: specifyweb/stored_queries/format.py
Comment (realVinayak) [reply specifyweb/stored_queries/format.py:126]: Yeah, except that you can detect cycles more definitely on the frontend rather than the backend. This is because frontend can check if the id is repeated. Backend cannot since the query hasn't been run yet.

**Comment by pashiav** [comment]
HOW DO WE TEST THIS PLEASE? </3 @realVinayak

**Comment by realVinayak** [comment]
**To-many not being aggregated when referenced in a formatter**
1. Open record formatter visual editor for a table.
2. Add a field which is to-many (for example, Determinations for Collection Object). It would appear as (aggregated)
3. If using Determinations for Collection Object, make a new collection object with more than 1 Determination.
4. Query that collection object, using (formatted) as the field. You should see that both of the determinations have been included in a single line.
Example:
image
My collection object has two determinations, one with Taxon named "Abantennarius will be added" and other with name "Aborichthys aa" and they both appear (since formatter of my determination is taxon)
image

**Review by bronwyncombs — CHANGES_REQUESTED** [review]
aggregator limits aren't enforced </3
Screenshot 2024-01-03 at 3 52 06 PM
Screenshot 2024-01-03 at 3 51 45 PM

**Review by CarolineDenis — APPROVED** [review]

**Review by bronwyncombs — CHANGES_REQUESTED** [review]
Probably shouldn't allow for users to set a formatter with an aggregated field, seems to make the software unhappy. 
Screenshot 2024-01-10 at 2 16 57 PM

Screenshot 2024-01-10 at 2 18 03 PM

**Comment by bronwyncombs** [comment]
Error not replicated on `xml-editor` branch

Screenshot 2024-01-10 at 2 55 20 PM
Screenshot 2024-01-10 at 2 53 49 PM

**Comment by bronwyncombs** [comment]
Recursive formatters still not handled by backend ([issue #4292](https://github.com/specify/specify7/issues/4292))

Screenshot 2024-01-10 at 3 23 14 PM
Screenshot 2024-01-10 at 3 22 55 PM

---

Deletion blockers give this error, not totally sure how this is tied to this PR but issue #3576 is linked so I'll comment on it

image

---

I applied the example given by Max in the description of issue #3266 with accession and accession agent. 

1. added accession agents (aggregated) to accession formatter
2. opened accession agent formatter
3. I received this error message as a toast

Screenshot 2024-01-10 at 3 34 04 PM

---

I also used the example given in the description of issue #3341 

This seems to be fixed as I only got one result from my query, but the videos and links wouldn't work on that issue so I am not positive

Screenshot 2024-01-10 at 3 50 11 PM
Screenshot 2024-01-10 at 3 52 31 PM

**Review by emenslin — CHANGES_REQUESTED** [review]
Still running into this issue that links to #3266 that Bronwyn mentioned above. 
Screenshot 2024-01-25 142245

**Comment by realVinayak** [comment]
@acwhite211 @bronwyncombs 

Hierarchy Exception is an unrelated [issue](https://github.com/specify/specify7/issues/3564).

**Review by emenslin — APPROVED** [review]
Looks good, it seems like everything that was supposed to be fixed has been fixed!

**Review by bronwyncombs — APPROVED** [review]
All issues raised that are within the scope of this PR have been resolved. Thanks for working on and improving this!

File: specifyweb/stored_queries/query_construct.py
Comment (melton-jason) [inline specifyweb/stored_queries/query_construct.py:23]: Outside of these two usages, where is `QueryConstruct.tree_rank_count` used? I can not find another use in the code base for the attribute.

**Review by melton-jason — APPROVED** [review]
At some point, we will definitely need to migrate away from the old `collections.namedtuple` class and at the very least update these classes (and hopefully others throughout the backend) to be subclasses of `typing.NamedTuple`. 

There was a lot to go through here, but the changes and added tests look okay. 

Further tests could be created testing ObjectFormatter methods `aggregate` and `objformat` when supplied with a specific aggregator/formatter name (besides None), but I would not consider that a necessity before this PR is merged.

File: specifyweb/stored_queries/query_construct.py
Comment (realVinayak) [reply specifyweb/stored_queries/query_construct.py:23]: Yeah, @acwhite211 you can rid of this. I was working on handling cases where formatter of taxon is defined with fields from the parent. In that case, the cycle will end (unlike other cyclical cases). Alec you can continue on that if you want!

**Review by acwhite211 — APPROVED** [review]

File: specifyweb/stored_queries/query_construct.py
Comment (acwhite211) [reply specifyweb/stored_queries/query_construct.py]: Sounds good, I'll keep it in, I added a TODO note in the code on implementing this in the future.

**Pull request merged**
